### ISSUE TITLE
The `Exec` task has an unmodeled dependency on the `$PATH` environment variable

### ISSUE THREAD

**Issue by rschmitt** [gradle/gradle#28110]
### Current Behavior

When configuring an `ExecSpec` or similar, it is not required to provide an actual _path_ to the executable; instead it is perfectly legal to supply an executable _name_, which will be resolved against the `$PATH` environment variable of the parent process. This resolution takes place in native code in the forked child process ([example](https://github.com/openjdk/jdk/blob/e5cb78cc88761cd27964e9fe77fc9c6f9073e888/src/java.base/unix/native/libjava/childproc.c#L277)), so it is completely invisible to Gradle and its instrumentation of methods like `System::getenv`. Changes to `$PATH` won't cause a configuration cache entry to be invalidated, for example:

```
Configuration cache entry stored.
[0] ws-BrazilGradle/src/SampleBrazilGradlePackage # PATH="$PATH:x" ./gradlew build --configuration-cache
Reusing configuration cache.

BUILD SUCCESSFUL in 491ms
24 actionable tasks: 4 executed, 20 up-to-date
Configuration cache entry reused.
```

This bug _might_ be benign: it's not clear that a `$PATH` change would ever require the task graph itself to change; the `Exec` task is already ineligible for caching by default; and according to the documentation, the Gradle client relays its environment variables to the Gradle daemon, which should work as long as the Gradle daemon actually mutates its own environment variables (i.e. through native code) instead of just faking it by intercepting `System::getenv` calls.

Regarding this last point: it's worth noting that the JDK's `ProcessBuilder` class internally does not go through `System::getenv` to get a copy of the environment (to use with the child process). It goes through a weird package-private class called `java.lang.ProcessEnvironment` which I doubt is instrumented by Gradle. This is worth looking into, to ensure that there really isn't a correctness issue here. I found this issue by inspection and I don't have a reproducer to share.

### Expected Behavior

The `$PATH` dependency should ideally be replaced with a modeled `Exec` task input, and Gradle should resolve the path to the executable itself, before calling into the JDK to spawn the process. 

### Steps to Reproduce

I can provide a reproducer upon request. First, I want to confirm that this isn't already a known issue.

### Gradle version

8.6

**Comment by lptr** [comment]
I agree that there is a dependency on the value of `PATH` which is unfortunate. However, even if we resolved the executable, it would not account for all inputs properly.

Here are some of the potential options of what we could do at this point:

1) Make the `Exec` task `@Untracked`. This would make it clear that Gradle cannot ensure that the same executable is used based on what `Exec` captures today (whatever "same executable" means exactly). If you wanted a cacheable/incremental task that executes a local command, you'd need to create your own task, relying on `ProcessOperations.exec()` in the task action, capturing your own inputs.
2) Capture the resolved absolute path of the executable; this would fix the reliance on `PATH` specifically. However, it would not ensure that, given different machines, or the same machines at different times, that the same executable would be used. It would also make the task be not cacheable across machines (unless they all had the same executable installed in the same location)
3) Instead of capturing the path, capture the hash of the executable binary. This would allow cross-machine caching, but only as long as you were on the same platform. Also, it would assume that the executable binary contained everything relevant for turning the input into the output. Specifically, if the executable relied on dynamic libraries, those would not be captured.

The point is, what "same executable" means depends on your specific use case, and Gradle cannot provide a general solution here. The current state of things is misleading, suggesting that `Exec` has enough information where it clearly doesn't. Therefore the best and safest option would be 1).

**Comment by digulla** [comment]
I have run into a corner case of this: I have a tool on my path. When I run Gradle and print `environment["PATH"]` in the Exec task, I can see that the folder on the path at configuration time. Yet, when I run `./gradlew`, I get `FILE_NOT_FOUND`. I tried a lot of solutions but so far, the only one that worked was to link the tool into $HOME/bin/ - forcing me to remember to switch the link when I work on different projects or an old version of the current project.

Looking at the discussion above, I see these options:

- Add a hint to Exec whether it's cacheable or not. The default should be "not". If the author of the project is confident, they can set this to true. This means Gradle doesn't have to guess or do a deep introspection which would fail in most cases. Just imagine, the tool is a script: Gradle would have to do a static analysis, find all the external programs mentioned, collect the hashes plus the  shared libraries being loaded ... for every script language that the OS supports...
- When Exec fails at runtime, it must print the path that was used.

The last step is problematic since the VM has some generous leeway how to implement the private class `ProcessImpl`. It might search the path, it might delegate this task to the OS. So even when we print the PATH passed to Gradle at config time, the runtime might still use something else. It would mean Gradle might have to determine the type and version of the VM it's running on to know which method is used ... per OS ...

My suggestion to fix this is add path searching to Gradle itself. The Exec task should be extended to accept a java.io.Path as "command name". It should remember the absolute path and pass that to the executor. Instead printing `foo: 2 file not found`, Gradle could print `/home/.../tool/foo/bin/foo.exe: 2 file not found`. With a config option, users could even enable a check at configuration time to fail early when the tool isn't on the path and Gradle could print "foo: Not found on PATH:" followed by a list of all entires, one per line.

If you supply just a string name, at configuration time, Gradle should examine the PATH variable and determine the absolute path of the executable itself and pass that as absolute java.io.Path. For remote builds, it should be possible to configure a list of java.io.Path items to search.

Advantages:
- It would make builds reliable
- It would allow to print good error messages
- It would allow Gradle to define the rules (instead of relying of some arcane knowledge about how PATH resolution works in Java, the OS, your shell...)
- It would give the writer of build scripts total control and ways to make their build work even when there are bugs in the first versions of this fix.
- If we expose the path searching algorithm as tool where each step is a protected method, they can patch it without having to invent all four wheels of the car again.

Disadvantages:
- It's an important change in a place where not a lot of people will expect it. The new code must cover all uncertainties with good error handling to guide users of Gradle towards solutions.
- Gradle path resolution would probably work slightly different than what experts are used to. I'm think about Windows experts and the case when FOO.BAT and FOO.EXE are both on the PATH or even in the same directory: Which one should Gradle pick? 

A simple first step to fix this would be to add a patch-searching tool so users can configure the Exec task with an absolute path of the executable at configuration time. When this works well, Exec can be changed to use this tool.

**Comment by digulla** [comment]
I've written a demo project to illustrate my idea: https://github.com/digulla/gradle-path-search-tool and would like to hear feedback.

**Comment by diegourban** [comment]
I have bumped into a similar issue and put together a self-contained MRE: 
https://github.com/diegourban/gradle-exec-path-mre (see README.md with steps to reproduce)

The trigger for me was IntelliJ on macOS: it started Gradle daemon with the minimal PATH. What I found is that `EstablishBuildEnvironment` patches `System.getenv("PATH")` in the JVM, but that change never reaches the OS-level `environ` array that native process spawning uses to resolve bare command names. 

So Exec tasks fail with "command not found" even though `System.getenv("PATH")` contains the right directory.
